utils/dvb/dvb-format-convert.1
.TH "dvb-format-convert" 1 "Fri Oct 3 2014" "DVBv5 Utils" "User Commands"
.SH NAME
dvb-format-convert \- DVBv5 tool for file format conversions
.SH SYNOPSIS
.B dvb-format-convert
[\fIOPTION\fR]...  \fIinput-file\fR \fIoutput-file\fR
.SH DESCRIPTION
dvb-format-convert is a tool meant to convert among different file formats.
It is compliant with version 5 of the DVB API, being capable of representing
all properties on any standard supported by the Linux digital TV drivers.
.PP
There are currently 3 different formats supported for input:
.IP "\(bu" 2
\fBdvbv5\fR \- the standard format at libdvbv5, capable of representing all
different TV standards;
.PP
.IP "\(bu" 2
\fBchannel\fR \- the dvb-apps legacy channel format, with supports only
ATSC, DVB-C, DVB-S and DVB-T standards, extended to also support s2-scan
format for DVB-S2 files, and to support DVB-T2.
.IP "\(bu" 2
\fBzap\fR \- the dvb-apps legacy format for tuning, with supports only
ATSC, DVB-C, DVB-S and DVB-T standards.
.PP
There is one extra output format:
.PP
.IP "\(bu" 2
\fBvdr\fR \- the format used by the vdr application. It supports all features
found up on vdr version 2.1.6.
.PP
Unfortunately, the dvb-apps legacy zap format doesn't store the DVB standard,
so, any conversions to and/or from this format requires an extra parameter,
to specify the delivery system.
.PP
.SH OPTIONS
.TP
The following options are valid:
.TP
\fB-I\fR, \fB--input-format\fR=\fIformat\fR
Format of the input file.
Supported input formats: \fBchannel\f, \fBzap\fR and \fBdvbv5\fR.
.TP
\fB-O\fR, \fB--output-format\fR=\fIformat\fR
Format of the output file.
Supported output formats: \fBvdr\fR, \fBchannel\fR, \fBzap\fR and \fBdvbv5\fR.
.TP
\fB-s\fR, \fB--delsys\fR=\fIsystem\fR
Delivery system type.
Needed if input or output format is ZAP.
.TP
\fB-?\fR, \fB--help\fR
Outputs the usage help.
.TP
\fB--usage\fR
Give a short usage message.
.TP
\fB-V\fR, \fB--version\fR
Print program version.
.SH EXAMPLES
.RS
.SS Converting a channel file from legacy dvb-apps format
Reads a file at the dvb\-apps channel file and writes it at dvbv5 format.
.PP
.nf
$ \fBdvb\-format\-convert \-I channel \-O dvbv5 dvbc\-channel\-legacy dvbc\-channel\fR
Reading file dvbc\-channel\-legacy
Writing file dvbc\-channel
.fi
.SS Convert a dvbv5 file with programs on it to the legacy dvb\-apps zap format
Reads a file generated by dvbv5-scan on dvbv5 format and writes a new file
at dvb\-apps zap format.
.PP
.nf
$ \fBdvb\-format\-convert \-I dvbv5 \-O zap \-s dvbc/annex_a dvb_channel.conf legacy.conf\fR
Reading file dvb_channel.conf
Writing file legacy.conf
WARNING: missing video PID while parsing entry 6 of legacy.conf
WARNING: missing video PID while parsing entry 7 of legacy.conf
WARNING: missing video PID while parsing entry 8 of legacy.conf
WARNING: missing audio PID while parsing entry 8 of legacy.conf
.fi
.PP
Please notice that the legacy dvb\-apps format assumes that every channel
will have one (and just one) audio PID and one (and just one) video PID,
while the dvbv5 format supports multiple PID per service type.
.PP
The above warnings happened on pure\-audio channels and on data\-only channels:
.PP
Original dvb_channel.conf file:
.PP
.nf
[the]
        SERVICE_ID = 4
        VIDEO_PID = 42
        AUDIO_PID = 257
        PID_f1 = 768
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[quick]
        SERVICE_ID = 5
        VIDEO_PID = 288
        AUDIO_PID = 289
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[brown]
        SERVICE_ID = 7
        VIDEO_PID = 36
        AUDIO_PID = 305
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[fox]
        SERVICE_ID = 13
        VIDEO_PID = 272
        AUDIO_PID = 273
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[jumps]
        SERVICE_ID = 18
        VIDEO_PID = 368
        AUDIO_PID = 369
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[over]
        SERVICE_ID = 48
        VIDEO_PID = 336
        AUDIO_PID = 337 338 849
        PID_86 = 816
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[the]
        SERVICE_ID = 57
        AUDIO_PID = 353
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[lazy]
        SERVICE_ID = 104
        AUDIO_PID = 321
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A


[dog]
        SERVICE_ID = 252
        PID_05 = 3047 3046 3045 3044 3042
        FREQUENCY = 573000000
        MODULATION = QAM/256
        INVERSION = AUTO
        SYMBOL_RATE = 5217000
        INNER_FEC = NONE
        DELIVERY_SYSTEM = DVBC/ANNEX_A

.fi
.PP
So, for the channels with multiple PIDs, the converted legacy.conf file
will have just the first one.
.PP
For the channels without an audio or video PID, it will store PID=0.
.PP
This is the output file for the above channel list.
.PP
.nf
the:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:42:257:4
quick:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:288:289:5
brown:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:36:305:7
fox:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:272:273:13
jumps:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:368:369:18
over:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:336:337:48
the:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:0:353:57
lazy:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:0:321:104
dog:573000000:INVERSION_AUTO:5217000:FEC_NONE:QAM_256:0:0:252
.fi
.SH BUGS
Report bugs to \fBLinux Media Mailing List <linux-media@vger.kernel.org>\fR
.SH COPYRIGHT
Copyright (c) 2011\-2014 by Mauro Carvalho Chehab.
.PP
License GPLv2: GNU GPL version 2 <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


ChangeLog
v4l-utils-1.5.91 (Development snapshot)
---------------------------------------

Gregor Jasny (3):
      Start v4l-utils 1.5.91 development cycle
      libdvbv5: Remove left over include of desc_service_list.h
      libdvbv5: Fix typo in doxygen index file

Mauro Carvalho Chehab (30):
      dvb-file: Fix a memory leak
      dvb-file: frees channel/vchannel if service id not found
      dvb-file: Remove spaces at the beginning of channel name
      dvb-file: Only print SDT warning if are there any channel to be stored
      dvb-file: Fix channel name when SDT is not found
      libdvbv5/descriptors: Fix content hexdump
      libdvbv5: Fix the number of frequencies at ISDB-T delivery sys descriptor
      libdvbv5: print ISDB-T frequencies in MHz
      libdvbv5: add a parser for ISO-10646/UCS2 strings
      libdvbv5: Fix ISDB-T TS info descriptor
      dvb-file: Fix no-SDT channel name allocation
      dvb-file: fix the memory deallocation on error
      dvb-file: don't dereference a NULL pointer
      libdvbv5: check for alloc errors at desc_ca
      dvb-format-convert: make valgrind happy
      libdvbv5: fix a nasty to track double-free bug
      dvb-fe-tool: abort nicely on femon mode
      libdvbv5: replace @include by @example
      man: move dvb-scan.1 out of Doxygen
      man: move dvb-fe-tool.1 out of Doxygen
      dvb-format-convert: Fix help message
      man: move dvb-format-convert.1 out of Doxygen
      man: add a utils/dvb/dvbv5-zap.1 page and remove from doxygen
      user man pages: Change report bugs to linux media ML
      Revert "dvb-file: don't dereference a NULL pointer"
      libdvbv5: Fix frequency table alloc size at desc_frequency_list
      .gitignore: Ignore quilt patches/ dir
      dvbv5-scan: add vdr format for the --help parameter
      user man pages: Add version to the man pages


v4l-utils-1.5.90 (Development snapshot)
---------------------------------------

André Roth (7):
      libdvbv5: cleanup logging, some memory checks
      libdvbv5: fix satellite LNBf handling
      libdvbv5: fix dvb_fe_dummy()
      libdvbv5: MPEG ES parser documentation
      libdvbv5: use hyperlinks in doxygen PDF
      libdvbv5: MPEG TS parser documentation
      libdvbv5: MPEG TS parser documentation and cleanups

Gregor Jasny (5):
      Start v4l-utils 1.5.0 development cycle
      buildsystem: Use latest ax_prog_doxygen from GNU Autoconf Archive
      libdvbv5: Add todo file
      libdvbv5: remove service_location descriptor
      libdvbv5: remove service_list descriptor
      doxygen: distribute doc/libdvbv5-index.doc in tarball

Hans Verkuil (9):
      v4l2-compliance: fix buffer state test
      (c)v4l-helpers: zero bytesperline, swap arg order
      qv4l2: the helper now zeroes bytesperline
      cv4l-helpers: add missing ioctls
      v4l2-ctl: fix control support for old kernels.
      v4l-helpers.h: fix control ioctls checks
      qv4l2: init dest format when starting vbi/swradio streaming
      v4l2-compliance: fix valgrind warnings
      qv4l2: fix typo: 576 instead of 575

Hans de Goede (3):
      libv4l2: Fix restoring of original dest-fmt after a VIDIOC_S_DV_TIMING
      libv4l2: Set convert_mmap_frame_size as soon as we've a dest_fmt
      libv4l2: Move alignment of dest_fmt resolution to v4l2_set_src_and_dest_format

Mauro Carvalho Chehab (60):
      libdvbv5: add doxygen tags for header.h
      libdvbv5: add doxygen tags to vct.h
      libdvbv5: add doxygen tags to pat.h
      doxygen_libdvbv5.cfg: add header.h, vct.h and pat.h
      libdvbv5: fix some cut-and-paste errors at documentation
      libdvbv5: add doxygen tags to pmt.h
      libdvbv5: better comment the descriptors
      libdvbv5: fix some issues at PAT documentation
      libdvbv5: add documentation for sdt.h
      libdvbv5: some documentation fixups for sdt.h
      libdvbv5: add documentation for the NIT table parser
      libdvbv5: add documentation for crc32.h
      libdvbv5: Add documentation for EIT DVB parser
      libdvbv5: Add a mainpage to the documentation
      libdvbv5: add documentation for ATSC EIT.
      libdvbv5: don't use ATSC_HEADER() macro
      libdvbv5: don't use DVB_DESC_HEADER() macro
      libdvbv5: add documentation for cat.h and atsc_header.h
      libdvbv5: add support for mgt.h
      libdvbv5 documentation review
      libdvbv5: group documentation into modules
      libdvbv5: add a missing macro to dvb_table group for documentation
      doxygen: add dvbv5-tools to the documentation
      dvb-fe-tool: Fix verbosity and comment unused option
      libdvbv5: fix a typo at ASTRA 1E extended LNBf
      Add support for VDR format (only for output)
      libdvbv5: Add support for opening FE on read only mode
      dvb-fe-tool: add support to open DVB FE in read only mode
      libdvbv5: Don't double-free the file name
      dvb-fe-tool: add femon parameter
      libdvbv5: fix a nasty bug when printing stats
      libdvbv5: Re-add the status name at the stats report
      libdvbv5: export dvb_fe_retrieve_quality()
      dvb-fe-tool: Colorize the status report
      dvb-fe-tool: implement acoustic mode
      update dvb-fe-tool Doxygen documentation
      libdvbv5: revert an ABI breakage
      libdvbv5: Create doxygen docs for mpeg_es.h
      libdvbv5: add documentation for ATSC service location
      libdvbv5: Add documentation for CA descriptor
      libdvbv5: Add documentation for CA identifier descriptor
      libdvbv5: add documentation for cable delivery system descriptor
      libdvbv5: add documentation for the extended event descriptor
      libdvbv5: add documentation for the short event descriptor
      libdvbv5: add documentation for the extended descriptor
      libdvbv5: add documentation for the frequency list descriptor
      libdvbv5: Document the hierarchy descriptor
      libdvbv5: Add documentation for ISDB-T terrestrial descriptor
      libdvbv5: add documentation for the language descriptor
      libdvbv5: Add description for the LCN descriptor

utils/dvb/dvb-format-convert.c
/*
 * Copyright (c) 2011-2012 - Mauro Carvalho Chehab
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation version 2
 * of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 * Or, point your browser to http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 *
 */

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>
#include <argp.h>

#include <config.h>

#include "libdvbv5/dvb-file.h"
#include "libdvbv5/dvb-demux.h"
#include "libdvbv5/dvb-scan.h"

#define PROGRAM_NAME	"dvb-format-convert"

struct arguments {
	char *input_file, *output_file;
	enum dvb_file_formats input_format, output_format;
	int delsys;
};

static const struct argp_option options[] = {
	{"input-format",	'I',	"format",	0, "Valid input formats: ZAP, CHANNEL, DVBV5", 0},
	{"output-format",	'O',	"format",	0, "Valid output formats: VDR, ZAP, CHANNEL, DVBV5", 0},
	{"delsys",		's',	"system",	0, "Delivery system type. Needed if input or output format is ZAP", 0},
	{ 0, 0, 0, 0, 0, 0 }
};

const char *argp_program_version = PROGRAM_NAME " version " V4L_UTILS_VERSION;
const char *argp_program_bug_address = "Mauro Carvalho Chehab <m.chehab@samsung.com>";

static error_t parse_opt(int k, char *optarg, struct argp_state *state)
{
	struct arguments *args = state->input;
	switch (k) {
	case 'I':
		args->input_format = dvb_parse_format(optarg);
		break;
	case 'O':
		args->output_format = dvb_parse_format(optarg);
		break;
	case 's':
		args->delsys = dvb_parse_delsys(optarg);
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	};
	return 0;
}

static int convert_file(struct arguments *args)
{
	struct dvb_file *dvb_file = NULL;
	int ret;

	printf("Reading file %s\n", args->input_file);

	dvb_file = dvb_read_file_format(args->input_file, args->delsys,
				    args->input_format);
	if (!dvb_file) {
		fprintf(stderr, "Error reading file %s\n", args->input_file);
		return -1;
	}

	printf("Writing file %s\n", args->output_file);
	ret = dvb_write_file_format(args->output_file, dvb_file,
				    args->delsys, args->output_format);
	dvb_file_free(dvb_file);

	return ret;
}

int main(int argc, char **argv)
{
	struct arguments args;
	int idx = -1, missing = 0;
	const struct argp argp = {
		.options = options,
		.parser = parse_opt,
		.doc = "scan DVB services using the channel file",
		.args_doc = "<input file> <output file>",
	};

	memset(&args, 0, sizeof(args));
	argp_parse(&argp, argc, argv, 0, &idx, &args);

	if (idx + 1 < argc) {
		args.input_file = argv[idx];
		args.output_file = argv[idx + 1];
	}

	if (args.input_format == FILE_UNKNOWN) {
		fprintf(stderr, "ERROR: Please specify a valid input format\n");
		missing = 1;
	} else if (!args.input_file) {
		fprintf(stderr, "ERROR: Please specify a valid input file\n");
		missing = 1;
	} else if (args.output_format == FILE_UNKNOWN) {
		fprintf(stderr, "ERROR: Please specify a valid output format\n");
		missing = 1;
	} else if (!args.output_file) {
		fprintf(stderr, "ERROR: Please specify a valid output file\n");
		missing = 1;
	} else if (((args.input_format == FILE_ZAP) ||
		   (args.output_format == FILE_ZAP)) && args.delsys <= 0) {
		fprintf(stderr, "ERROR: Please specify a valid delivery system for ZAP format\n");
		missing = 1;
	}
	if (missing) {
		argp_help(&argp, stderr, ARGP_HELP_STD_HELP, PROGRAM_NAME);
		return -1;
	}

	return convert_file(&args);
}
